### PR TITLE
run: Add cgroup-parent flag

### PR DESCRIPTION
**Pull request by dcantah**
Adds a test case and the functionality for cgroup-parent to match docker runs support.

### CODE REVIEW

File: cmd/nerdctl/run_cgroup_linux.go [hunk 1]
```diff
@@ -86,28 +86,31 @@ func generateCgroupOpts(cmd *cobra.Command, id string) ([]oci.SpecOpts, error) {
...
 		if !rootlessutil.IsRootless() {
 			return nil, errors.New("cgroup-manager \"none\" is only supported for rootless")
 		}
 
-		if cpus > 0.0 || memStr != "" || memSwap != "" || pidsLimit > 0 {
-			logrus.Warn("cgroup manager is set to \"none\", discarding resource limit requests. " +
+		if cpus > 0.0 || memStr != "" || memSwap != "" || pidsLimit > 0 || parent != "" {
+			logrus.Warn(`cgroup manager is set to "none", discarding resource limit requests and ignoring cgroup parent if specified. ` +
 				"(Hint: enable cgroup v2 with systemd: https://rootlesscontaine.rs/getting-started/common/cgroup2/)")
 		}
```
Comment (AkihiroSuda) [inline cmd/nerdctl/run_cgroup_linux.go:103]:
```suggestion
		if cpus > 0.0 || memStr != "" || memSwap != "" || pidsLimit > 0 {
			logrus.Warn(`cgroup manager is set to "none", discarding resource limit requests. ` +
				"(Hint: enable cgroup v2 with systemd: https://rootlesscontaine.rs/getting-started/common/cgroup2/)")
		}
				if parent != "" {
			logrus.Warng(`cgroup manager is set to "none", ignoring cgroup parent %q` +
				"(Hint: enable cgroup v2 with systemd: https://rootlesscontaine.rs/getting-started/common/cgroup2/)", parent)
		}
```

Comment (dcantah) [reply cmd/nerdctl/run_cgroup_linux.go:103]: @AkihiroSuda Done!

File: cmd/nerdctl/run_cgroup_linux.go [hunk 2]
```diff
@@ -86,28 +86,35 @@ func generateCgroupOpts(cmd *cobra.Command, id string) ([]oci.SpecOpts, error) {
...
+
 	if cgroupManager == "none" {
 		if !rootlessutil.IsRootless() {
-			return nil, errors.New("cgroup-manager \"none\" is only supported for rootless")
+			return nil, errors.New(`cgroup-manager "none" is only supported for rootless`)
 		}
 
 		if cpus > 0.0 || memStr != "" || memSwap != "" || pidsLimit > 0 {
-			logrus.Warn("cgroup manager is set to \"none\", discarding resource limit requests. " +
+			logrus.Warn(`cgroup manager is set to "none", discarding resource limit requests. ` +
```
Comment (AkihiroSuda) [inline cmd/nerdctl/run_cgroup_linux.go:101]: Any reason for this change?

File: cmd/nerdctl/run_cgroup_linux.go [hunk 3]
```diff
@@ -86,28 +86,35 @@ func generateCgroupOpts(cmd *cobra.Command, id string) ([]oci.SpecOpts, error) {
...
+
+	parent, err := cmd.Flags().GetString("cgroup-parent")
+	if err != nil {
+		return nil, err
+	}
+
 	if cgroupManager == "none" {
 		if !rootlessutil.IsRootless() {
-			return nil, errors.New("cgroup-manager \"none\" is only supported for rootless")
+			return nil, errors.New(`cgroup-manager "none" is only supported for rootless`)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/run_cgroup_linux.go:97]: Any reason for this change?

File: cmd/nerdctl/run_cgroup_linux_test.go [hunk 4]
```diff
@@ -283,6 +283,50 @@ func TestRunCgroupConf(t *testing.T) {
...
+		"run",
+		"-d",
+		"--name",
+		containerName,
+		"--cgroupns=host",
+		"--cgroup-parent", parent,
+		testutil.AlpineImage,
+		"sleep",
+		"infinity",
+	)
```
Comment (manugupt1) [inline cmd/nerdctl/run_cgroup_linux_test.go:319]: base.Cmd(...).AssertOK() should do the trick

File: cmd/nerdctl/run_cgroup_linux_test.go [hunk 5]
```diff
@@ -283,6 +283,50 @@ func TestRunCgroupConf(t *testing.T) {
...
+	cmd.Run()
+
+	id := cmd.Base.InspectContainer(containerName).ID
```
Comment (manugupt1) [inline cmd/nerdctl/run_cgroup_linux_test.go:322]: base.InspectContainer?

File: cmd/nerdctl/run_cgroup_linux_test.go [hunk 6]
```diff
@@ -283,6 +283,50 @@ func TestRunCgroupConf(t *testing.T) {
...
+	containerName := testutil.Identifier(t)
+	defer base.Cmd("rm", "-f", containerName).Run()
+
+	switch info.CgroupDriver {
+	case "none", "":
+		t.Skip("test requires cgroup driver")
+	}
+
+	parent := "/foobarbaz"
+	if info.CgroupDriver == "systemd" {
```
Comment (manugupt1) [inline cmd/nerdctl/run_cgroup_linux_test.go:299]: Can you log out somewhere which cgroup drivers are being run on the github host to understand which path are we testing?

Comment (dcantah) [reply cmd/nerdctl/run_cgroup_linux_test.go:299]: Sure, and I'll look/address the other feedback sometime this weekend. Thanks!

File: cmd/nerdctl/run_cgroup_linux_test.go [hunk 7]
```diff
@@ -283,6 +283,50 @@ func TestRunCgroupConf(t *testing.T) {
 		"cat", "memory.high").AssertOutExactly("33554432\n")
 }
 
+func TestRunCgroupParent(t *testing.T) {
+	t.Parallel()
+	base := testutil.NewBase(t)
+	info := base.Info()
+	containerName := testutil.Identifier(t)
+	defer base.Cmd("rm", "-f", containerName).Run()
```
[the remark this answers is not in the document]
Comment (dcantah) [reply cmd/nerdctl/run_cgroup_linux_test.go:291]: Are you referring to the deferred cleanup? I was just following the standard set in most of the other tests in this file, but seems it's actually the lesser used of the two when taking into account all of the other tests; only run_cgroup_linux_test and run_mount_linux_test just defer `Run` and call it a day. I'll open up a PR to swap the cgroup tests to using `AssertOK` instead (and use `AssertOK` here

File: cmd/nerdctl/run_cgroup_linux_test.go (hunk 5 above)
Comment (dcantah) [reply cmd/nerdctl/run_cgroup_linux_test.go:322]: Yep, no need to chain off cmd, will change

File: cmd/nerdctl/run_cgroup_linux_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (dcantah) [reply cmd/nerdctl/run_cgroup_linux_test.go:291]: https://github.com/containerd/nerdctl/pull/1789

File: cmd/nerdctl/run_cgroup_linux_test.go (hunk 6 above)
Comment (dcantah) [reply cmd/nerdctl/run_cgroup_linux_test.go:301]: Added in this test

File: cmd/nerdctl/run_cgroup_linux_test.go (hunk 5 above)
Comment (dcantah) [reply cmd/nerdctl/run_cgroup_linux_test.go]: Fixed

File: cmd/nerdctl/run_cgroup_linux_test.go (hunk 4 above)
Comment (dcantah) [reply cmd/nerdctl/run_cgroup_linux_test.go]: Done!

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**